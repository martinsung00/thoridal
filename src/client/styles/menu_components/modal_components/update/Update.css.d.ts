export const mainLogo: string;
export const notFound: string;
export const notFoundTag: string;
export const notFoundState: string;
export const confirmState: string;
export const confirm: string;
export const confirmTag: string;
