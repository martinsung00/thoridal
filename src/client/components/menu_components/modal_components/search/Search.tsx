import React from "react";
import axios, { AxiosResponse } from "axios";
import styles from "../../../../styles/menu_components/modal_components/search/Search.css";
import { Trade } from "../../../../types/index";
import { Trades } from "./Trades";

interface Props {
  toggleShow: Function;
  show: boolean;
}

interface State {
  current_page: "search" | "confirm" | "failed";
  errored: boolean;
  searchTerm: string;
  searchType: string;
  trades: Array<Trade> | [];
  page: number;
  pages: number;
}

export default class ModalContent extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      current_page: "search",
      errored: false,
      searchTerm: "",
      searchType: "id",
      trades: [],
      page: 1,
      pages: 0,
    };
  }

  componentDidMount(): void {
    const component = this;

    axios
      .get("/trade/all/find")
      .then(function (response) {
        const data: Array<Trade> = response.data.rows;

        component.setState(
          {
            trades: data.reverse(),
            errored: false,
          },
          function () {
            const pages = Math.ceil(component.state.trades.length / 10);

            component.setState({
              pages: pages,
            });
          }
        );
      })
      .catch(function (error: Error) {
        console.error(error);
        component._handleError();
      });
  }

  public _handleSelect(event: React.ChangeEvent<HTMLSelectElement>): void {
    event.preventDefault();

    this.setState({
      searchType: event.currentTarget.value,
    });
  }

  public _handleChange(event: React.ChangeEvent<HTMLInputElement>): void {
    event.preventDefault();

    this.setState({
      searchTerm: event.target.value,
    });
  }

  public _handleDelete(
    event: React.MouseEvent<HTMLDivElement, MouseEvent>,
    index: number,
    id: string
  ): void {
    event.preventDefault();

    const component = this;

    axios
      .delete(`/trade/id/${id}/delete`)
      .then(function () {
        component.state.trades.splice(index, 1);
        const data = component.state.trades;
        component.setState({
          trades: data,
        });
      })
      .catch(function (error: Error) {
        console.error(error);
        component._handleError();
      });
  }

  public _handleError(): void {
    this.setState({
      errored: true,
    });
  }

  public _handleSearch(event: React.FormEvent): void {
    event.preventDefault();

    const component = this;
    const type = this.state.searchType;
    const term = this.state.searchTerm;

    axios
      .get(`/trade/${type}/${term}/find`)
      .then(function (response: AxiosResponse) {
        const data: Array<Trade> = response.data.rows;
        component.setState({
          trades: data,
          errored: false,
        });
      })
      .catch(function (error) {
        console.error(error);
        component._handleError();
      });
  }

  render() {
    return (
      <div>
        <div>
          <input
            placeholder="Search Here!"
            required
            className={styles.searchBar}
            id="search-by"
            value={this.state.searchTerm}
            onChange={this._handleChange.bind(this)}
          ></input>
          <div className={styles.searchType}>
            <select
              className={styles.searchBy}
              value={this.state.searchType}
              defaultValue="id"
              onChange={this._handleSelect.bind(this)}
            >
              <option value="id">ID</option>
              <option value="ticker">Ticker</option>
              <option value="company">Company</option>
              <option value="date">Date</option>
            </select>
          </div>
          {this.state.searchTerm.length > 0 && (
            <div
              className={styles.searchButton}
              onClick={this._handleSearch.bind(this)}
            >
              Search
            </div>
          )}
          {this.state.searchTerm.length <= 0 && (
            <div className={styles.searchButtonGray}>Search</div>
          )}
        </div>
        <div className={styles.divider} />
        {this.state.trades.length > 0 && (
          <Trades
            page={this.state.page}
            pages={this.state.pages}
            trades={this.state.trades}
            handleDelete={this._handleDelete.bind(this)}
          />
        )}
        {this.state.trades.length === 0 && (
          <div className={styles.noHistory}>No Previous Trade History</div>
        )}
      </div>
    );
  }
}
