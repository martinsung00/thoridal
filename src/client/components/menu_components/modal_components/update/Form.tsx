import React from "react";
import axios from "axios";
import styles from "../../../../styles/menu_components/modal_components/update/UpdateForm.css";
import { Trade } from "../../../../types/index";

interface Props {
  onError: Function;
  nextState: Function;
  handleClose: Function;
  trades: Array<Trade>;
}

interface State {
  id: string;
  ticker: string;
  company_name: string;
  reference_number: string;
  unit_price: number;
  quantity: number;
  total_cost: number;
  trade_type: string;
  note: string;
  created_at: Date;
  trade_status: boolean;
  trade_action: string;
}

export default class Form extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    const prevTrade = this.props.trades[0];

    this.state = {
      id: prevTrade.id,
      ticker: prevTrade.ticker,
      company_name: prevTrade.company_name,
      reference_number: prevTrade.reference_number,
      unit_price: prevTrade.unit_price,
      quantity: prevTrade.quantity,
      total_cost: prevTrade.total_cost,
      trade_type: prevTrade.trade_type,
      note: prevTrade.note,
      created_at: prevTrade.created_at,
      trade_status: prevTrade.trade_status,
      trade_action: prevTrade.trade_action,
    };
  }

  public _handleSelect(event: React.FormEvent<HTMLSelectElement>): void {
    event.preventDefault();

    this.setState({
      trade_type: event.currentTarget.value,
    });
  }

  public _handleTradeAction(event: React.FormEvent<HTMLSelectElement>): void {
    event.preventDefault();

    this.setState({
      trade_action: event.currentTarget.value,
    });
  }

  public _handleTradeStatus(event: React.FormEvent<HTMLSelectElement>): void {
    event.preventDefault();

    if (event.currentTarget.value === "fulfilled") {
      this.setState({
        trade_status: true,
      });
    } else {
      this.setState({
        trade_status: false,
      });
    }
  }

  public _handleChange(event: React.ChangeEvent<HTMLInputElement>): void {
    const key = event.target.name;
    const value = event.target.value;

    event.preventDefault();

    if (Object.keys(this.state).includes(key)) {
      this.setState({
        ...this.state,
        [key]: value,
      });
    }
  }

  public _onSubmit(event: React.FormEvent<HTMLFormElement>): void {
    event.preventDefault();

    const component = this;
    const document: Trade = {
      id: this.state.id,
      ticker: this.state.ticker,
      company_name: this.state.company_name,
      reference_number: this.state.reference_number,
      unit_price: this.state.unit_price,
      quantity: this.state.quantity,
      total_cost: this.state.total_cost,
      trade_type: this.state.trade_type === "short" ? "short" : "long",
      note: this.state.note,
      created_at: new Date(),
      trade_status: this.state.trade_status,
      trade_action: this.state.trade_action === "sold" ? "sold" : "bought",
    };

    axios
      .put("/trade/user/write", document)
      .catch(function (error) {
        component.props.onError();
        throw error;
      })
      .then(function () {
        component.props.nextState();
      });
  }

  render() {
    const component = this;

    return (
      <div className={styles.form}>
        <form onSubmit={this._onSubmit.bind(this)} className="inputForm">
          <label className={styles.label}>
            ID:{"  "}
            <input
              disabled
              type="string"
              placeholder="e.g 1234ABC"
              required
              name="id"
              id="id"
              value={this.state.id}
              className={styles.inputBoxID}
              onChange={this._handleChange.bind(this)}
            />
          </label>
          <div className={styles.spacing} />
          <label className={styles.label}>
            Company Name:{"  "}
            <input
              type="string"
              placeholder="e.g abc"
              required
              name="company_name"
              id="company_name"
              value={this.state.company_name}
              className={styles.inputBoxCompany}
              onChange={this._handleChange.bind(this)}
            />
          </label>
          <div className={styles.spacing} />
          <label className={styles.label}>
            Ticker:{"  "}
            <input
              type="string"
              placeholder="e.g abc"
              required
              name="ticker"
              id="ticker"
              value={this.state.ticker}
              className={styles.inputBoxTicker}
              onChange={this._handleChange.bind(this)}
            />
          </label>
          <div className={styles.spacing} />
          <label className={styles.label}>
            Reference Number:{"  "}
            <input
              type="string"
              placeholder="e.g 1234ABC"
              required
              name="reference_number"
              id="reference_number"
              value={this.state.reference_number}
              className={styles.inputBoxRef}
              onChange={this._handleChange.bind(this)}
            />
          </label>
          <div className={styles.spacing} />
          <label className={styles.label}>
            Unit Price:{"  "}
            <input
              type="number"
              placeholder="e.g 19.99"
              required
              name="unit_price"
              id="unit_price"
              value={this.state.unit_price}
              className={styles.inputBoxUnitPrice}
              onChange={this._handleChange.bind(this)}
            />
          </label>
          <div className={styles.spacing} />
          <label className={styles.label}>
            Quantity:{"  "}
            <input
              type="number"
              placeholder="e.g 100"
              required
              name="quantity"
              id="quantity"
              value={this.state.quantity}
              className={styles.inputBoxQuantity}
              onChange={this._handleChange.bind(this)}
            />
          </label>
          <div className={styles.spacing} />
          <label className={styles.label}>
            Total Cost:{"  "}
            <input
              type="number"
              placeholder="e.g 199.99"
              required
              name="total_cost"
              id="total_cost"
              value={this.state.total_cost}
              className={styles.inputBoxTotal}
              onChange={this._handleChange.bind(this)}
            />
          </label>
          <div className={styles.spacing} />
          <label className={styles.label}>
            Trade Type:{"  "}
            <select
              onChange={this._handleSelect.bind(this)}
              value={this.state.trade_type}
              defaultValue="short"
              id="trade_type"
              className={styles.selectBoxType}
            >
              <option value="long">Long</option>
              <option value="short">Short</option>
            </select>
          </label>
          <div className={styles.spacing} />
          <label className={styles.label}>
            Trade Status:{"  "}
            <select
              onChange={this._handleTradeStatus.bind(this)}
              value={
                this.state.trade_status === true ? "fulfilled" : "not-fulfilled"
              }
              defaultValue="not-fulfilled"
              id="trade_status"
              className={styles.selectBoxStatus}
            >
              <option value="fulfilled">Fulfilled</option>
              <option value="not-fulfilled">Not Fulfilled</option>
            </select>
          </label>
          <div className={styles.spacing} />
          <label className={styles.label}>
            Trade Action:{"  "}
            <select
              onChange={this._handleTradeAction.bind(this)}
              value={this.state.trade_action}
              id="trade_action"
              className={styles.selectBoxAction}
            >
              <option value="bought">Bought</option>
              <option value="sold">Sold</option>
            </select>
          </label>
          <div className={styles.spacing} />
          <label className={styles.label}>
            Note:{"  "}
            <input
              type="string"
              placeholder="e.g Hello World"
              required
              name="note"
              id="note"
              value={this.state.note}
              className={styles.inputBoxNote}
              onChange={this._handleChange.bind(this)}
            />
          </label>
          <div className={styles.spacing} />
          <input className={styles.submit} type="submit" value="Submit" />
        </form>
        <div
          className={styles.cancel}
          id="cancel-button"
          onClick={function (
            event: React.MouseEvent<HTMLDivElement, MouseEvent>
          ): void {
            event.preventDefault();

            component.props.handleClose(event);
          }}
        >
          <div className={styles.cancelText}>Cancel</div>
        </div>
      </div>
    );
  }
}
