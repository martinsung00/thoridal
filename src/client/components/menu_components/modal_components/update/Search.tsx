import React from "react";
import styles from "../../../../styles/menu_components/modal_components/update/UpdateSearch.css";

interface Props {
  // To do: Add props as needed
  handleChange: Function;
  handleSelect: Function;
  handleSearch: Function;
  updateBy: string;
}

export const Search: React.FunctionComponent<Props> = (props: Props) => {
  return (
    <div className={styles.root}>
      <div>Update Trade By:</div>
      <select
        className={styles.searchBy}
        defaultValue="id"
        id="update-by"
        value={props.updateBy}
        onChange={function (event: React.ChangeEvent<HTMLSelectElement>) {
          props.handleSelect(event);
        }}
      >
        <option value="id">ID</option>
        <option value="ticker">Ticker</option>
        <option value="company">Company Name</option>
        <option value="date">Date</option>
      </select>
      <input
        placeholder={props.updateBy === "date" ? "MM-DD-YYYY" : ""}
        required
        className={styles.searchBar}
        id="update-input"
        onChange={function (event: React.ChangeEvent<HTMLInputElement>) {
          props.handleChange(event);
        }}
      ></input>
      <div
        className={styles.searchButton}
        id="search-button"
        onClick={function (
          event: React.MouseEvent<HTMLDivElement, MouseEvent>
        ) {
          props.handleSearch(event);
        }}
      >
        <div className={styles.searchButtonTag}>Search</div>
      </div>
    </div>
  );
};
