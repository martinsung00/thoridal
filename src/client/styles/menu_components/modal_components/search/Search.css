.mainLogo {
  position: relative;
  height: 35px;
  width: 35px;
  margin-left: 47.5%;
}

.searchBar {
  position: relative;
  width: 20em;
  height: 1em;
  font-size: 22px;
  font-family: Helvetica, Arial, sans-serif;
  top: 3em;
  left: 26%;
  border: 3px solid gray;
  border-radius: 3px;
}

.searchType {
  position: relative;
  margin-right: 36em;
  margin-top: 4.8em;
  float: right;
  background-color: rgba(255, 255, 255, 0.952);
  padding-left: 3%;
  padding-right: -1%;
  border-left: 0.7px solid darkgray;
  width: 2.5em;
  font-family: Helvetica, Arial, sans-serif;
  font-size: 15px;
  outline: none;
}

.searchBy {
  border: none;
  outline: none;
  font-size: 14.5px;
  font-weight: bold;
  color: rgb(116, 116, 116);
  background-color: white;
  cursor: pointer;
}

.searchButton {
  position: relative;
  text-align: center;
  padding-bottom: 0.25%;
  left: 19.5%;
  float: right;
  width: 4em;
  height: 1em;
  font-size: 22px;
  font-family: Helvetica, Arial, sans-serif;
  top: 3em;
  color: white;
  background-color: black;
  border: 3px solid black;
  border-radius: 3px;
  cursor: pointer;
}

.searchButtonGray {
  position: relative;
  text-align: center;
  padding-bottom: 0.25%;
  left: 19.5%;
  float: right;
  width: 4em;
  height: 1em;
  font-size: 22px;
  font-family: Helvetica, Arial, sans-serif;
  top: 3em;
  color: rgb(0, 0, 0);
  background-color: rgb(99, 99, 99);
  border: 3px solid rgb(100, 100, 100);
  border-radius: 3px;
  cursor: pointer;
}

.searchButton:hover {
  position: relative;
  text-align: center;
  padding-bottom: 0.25%;
  left: 19.5%;
  float: right;
  width: 4em;
  height: 1em;
  font-size: 22px;
  font-family: Helvetica, Arial, sans-serif;
  top: 3em;
  color: rgb(0, 0, 0);
  background-color: rgb(180, 180, 180);
  border: 3px solid rgb(180, 180, 180);
  border-radius: 3px;
  cursor: pointer;
}

.divider {
  position: relative;
  height: 2em;
  top: 9em;
  left: 3.7em;
  width: 90%;
  position: relative;
  text-align: center;
  border-top: 0.7px solid black;
  border-bottom: 0.7px solid black;
}

.noHistory {
  position: relative;
  top: 5em;
  text-align: center;
  font-size: 35px;
  font-family: Helvetica, Arial, sans-serif;
  margin-top: 2em;
}
