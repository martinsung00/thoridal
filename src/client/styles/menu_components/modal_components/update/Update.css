.mainLogo {
  position: relative;
  height: 35px;
  width: 35px;
  margin-left: 47.5%;
}

.notFound {
  position: relative;
  float: left;
  top: 5em;
  left: 16em;
  height: 1.85em;
  width: 6em;
  background-color: rgb(0, 0, 0);
  color: rgb(255, 255, 255);
  font-size: 20px;
  text-align: center;
  border-radius: 3px;
  border: 1.3px solid rgb(0, 0, 0);
  cursor: pointer;
}

.notFound:hover {
  position: relative;
  float: left;
  top: 5em;
  left: 16em;
  height: 1.85em;
  width: 6em;
  background-color: darkgray;
  color: rgb(0, 0, 0);
  font-size: 20px;
  text-align: center;
  border-radius: 3px;
  border: 1.3px solid black;
  cursor: pointer;
  transition: ease-out 0.15s;
}

.notFoundTag {
  position: relative;
  margin-top: 0.4em;
}

.notFoundState {
  position: relative;
  font-size: 30px;
  font-family: Helvetica, Arial, sans-serif;
  text-align: center;
  color: black;
  top: 5em;
  left: 0;
}

.confirmState {
  position: relative;
  font-size: 30px;
  font-family: Helvetica, Arial, sans-serif;
  text-align: center;
  color: black;
  top: 5em;
  left: 0;
}

.confirm {
  position: relative;
  top: 5em;
  left: 16em;
  height: 1.85em;
  width: 6em;
  background-color: black;
  color: rgb(255, 255, 255);
  font-size: 20px;
  text-align: center;
  border-radius: 3px;
  border: 1.3px solid darkgray;
  cursor: pointer;
}

.confirm:hover {
  position: relative;
  top: 5em;
  left: 16em;
  height: 1.85em;
  width: 6em;
  background-color: rgb(255, 255, 255);
  color: rgb(0, 0, 0);
  font-size: 20px;
  text-align: center;
  border-radius: 3px;
  border: 1.3px solid black;
  cursor: pointer;
  transition: ease-out 0.15s;
}

.confirmTag {
  position: relative;
  margin-top: 0.4em;
}
