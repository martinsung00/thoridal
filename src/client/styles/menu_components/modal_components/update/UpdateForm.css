.mainLogo {
  position: relative;
  height: 35px;
  width: 35px;
  margin-left: 47.5%;
}

.spacing {
  padding-bottom: 1.5em;
}

.form {
  width: 100%;
  position: absolute;
  margin-top: 4%;
  margin-left: 26%;
  font-family: Helvetica, Arial, sans-serif;
  text-align: left;
}

.label {
  position: relative;
  font-family: Helvetica, Arial, sans-serif;
  font-size: 18px;
  padding-bottom: 5%;
}

.inputBoxID {
  position: relative;
  width: 10em;
  outline: none;
  border-top: none;
  border-left: none;
  border-right: none;
  border-bottom: 1.5px solid rgb(0, 0, 0);
  font-size: 18px;
  margin-left: 20%;
}

.inputBoxCompany {
  position: relative;
  width: 10em;
  outline: none;
  border-top: none;
  border-left: none;
  border-right: none;
  border-bottom: 1.5px solid rgb(0, 0, 0);
  font-size: 18px;
  margin-left: 6.1%;
  background-color: none;
}

.inputBoxTicker {
  position: relative;
  width: 10em;
  outline: none;
  border-top: none;
  border-left: none;
  border-right: none;
  border-bottom: 1.5px solid rgb(0, 0, 0);
  font-size: 18px;
  margin-left: 16.25%;
  background-color: none;
}

.inputBoxRef {
  position: relative;
  width: 10em;
  outline: none;
  border-top: none;
  border-left: none;
  border-right: none;
  border-bottom: 1.5px solid rgb(0, 0, 0);
  font-size: 18px;
  margin-left: 3.42%;
}

.inputBoxUnitPrice {
  position: relative;
  width: 10em;
  outline: none;
  border-top: none;
  border-left: none;
  border-right: none;
  border-bottom: 1.5px solid rgb(0, 0, 0);
  font-size: 18px;
  margin-left: 12.57%;
}

.inputBoxQuantity {
  position: relative;
  width: 10em;
  outline: none;
  border-top: none;
  border-left: none;
  border-right: none;
  border-bottom: 1.5px solid rgb(0, 0, 0);
  font-size: 18px;
  margin-left: 14%;
}

.inputBoxTotal {
  position: relative;
  width: 10em;
  outline: none;
  border-top: none;
  border-left: none;
  border-right: none;
  border-bottom: 1.5px solid rgb(0, 0, 0);
  font-size: 18px;
  margin-left: 12.4%;
}

.inputBoxNote {
  position: relative;
  margin-top: -5%;
  width: 10em;
  outline: none;
  border-top: none;
  border-left: none;
  border-right: none;
  border-bottom: none;
  font-size: 18px;
  margin-left: 17.4%;
  word-wrap: break-word;
}

.selectBoxType {
  position: relative;
  outline: none;
  border: none;
  font-size: 18px;
  font-family: Helvetica, Arial, sans-serif;
  margin-left: 10.5%;
  cursor: pointer;
}

.selectBoxStatus {
  position: relative;
  outline: none;
  border: none;
  font-size: 18px;
  font-family: Helvetica, Arial, sans-serif;
  margin-left: 9%;
  cursor: pointer;
}

.selectBoxAction {
  position: relative;
  outline: none;
  border: none;
  font-size: 18px;
  font-family: Helvetica, Arial, sans-serif;
  margin-left: 9.3%;
  cursor: pointer;
}

.submit {
  position: relative;
  height: 2em;
  width: 6em;
  background-color: black;
  border-radius: 3px;
  font-family: Helvetica, Arial, sans-serif;
  font-size: 20px;
  color: white;
  margin-top: 0%;
  margin-left: 25%;
  text-align: center;
}

.submit:hover {
  position: relative;
  height: 2em;
  width: 6em;
  background-color: darkgray;
  border-radius: 3px;
  font-family: Helvetica, Arial, sans-serif;
  font-size: 20px;
  color: black;
  margin-top: 0%;
  margin-left: 25%;
  text-align: center;
  transition: ease-out 0.15s;
}

.cancel {
  position: relative;
  float: left;
  height: 1.85em;
  width: 6em;
  background-color: rgb(255, 255, 255);
  border: 1.3px solid black;
  border-radius: 3px;
  font-family: Helvetica, Arial, sans-serif;
  font-size: 20px;
  color: black;
  margin-top: -4.95%;
  margin-left: 2%;
  text-align: center;
}

.cancel:hover {
  position: relative;
  float: left;
  height: 1.85em;
  width: 6em;
  background-color: darkgray;
  border-radius: 3px;
  font-family: Helvetica, Arial, sans-serif;
  font-size: 20px;
  color: black;
  margin-top: -4.95%;
  margin-left: 2%;
  text-align: center;
  transition: ease-out 0.15s;
}

.cancelText {
  position: relative;
  margin-top: 6.3%;
  right: -0.5%;
}
