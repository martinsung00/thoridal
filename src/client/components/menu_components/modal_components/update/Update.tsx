import React from "react";
import axios, { AxiosResponse } from "axios";
import styles from "../../../../styles/menu_components/modal_components/update/Update.css";
import Form from "./Form";
import { Search } from "./Search";
import { ErrorPage } from "../../../error_components/Error";
import { NotFoundPage } from "../../../error_components/NotFound";
import { Trade } from "../../../../types/index";

interface Props {
  // Add props when necessary
  toggleShow: Function;
  show: boolean;
}

interface State {
  current_page:
    | "find"
    | "failed"
    | "update-form"
    | "not-found"
    | "failed"
    | "confirm";
  errored: boolean;
  trades: Array<Trade>;
  searchTerm: string;
  updateBy: string;
}

export default class ModalContent extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      current_page: "find",
      errored: false,
      trades: [],
      searchTerm: "",
      updateBy: "id",
    };
  }

  // Intelligently proceed to the next correct page. If somehow current page is altered without permission, it will default to error page
  public _nextState(): void {
    if (this.state.current_page === "find") {
      this.setState({
        current_page: "update-form",
      });
    } else if (this.state.current_page === "update-form") {
      this.setState({
        current_page: "confirm",
      });
    } else {
      this._handleError();
    }
  }

  public _prevState(event: React.MouseEvent<HTMLDivElement, MouseEvent>) {
    event.preventDefault();

    this.setState({
      errored: false,
      current_page: "find",
    });
  }

  public _handleError(): void {
    this.setState({
      errored: true,
      current_page: "failed",
    });
  }

  public _handleNotFound(): void {
    this.setState({
      current_page: "not-found",
    });
  }

  public _handleClose(
    event: React.MouseEvent<HTMLDivElement, MouseEvent>
  ): void {
    event.preventDefault();
    this.setState({
      current_page: "find",
    });
    this.props.toggleShow(!this.props.show);
  }

  public _handleChange(event: React.ChangeEvent<HTMLInputElement>): void {
    event.preventDefault();

    this.setState({
      searchTerm: event.target.value,
    });
  }

  public _handleSelect(event: React.ChangeEvent<HTMLSelectElement>): void {
    event.preventDefault();

    this.setState({
      updateBy: event.currentTarget.value,
    });
  }

  public _handleSearch(
    event: React.MouseEvent<HTMLDivElement, MouseEvent>
  ): void {
    event.preventDefault();

    const component = this;
    const query = this.state.searchTerm;
    const type = this.state.updateBy;

    // If no input was made and submit is pressed
    if (!query.length || query.length === 0) {
      component._handleNotFound();
      return;
    }

    axios
      .get(`/trade/${type}/${query}/find`)
      .then(function (response: AxiosResponse) {
        const data: Array<Trade> = response.data.rows;

        if (!data.length || data.length === 0) {
          component._handleNotFound();
        } else {
          component.setState({
            trades: data,
          });
          component._nextState();
        }
      })
      .catch(function (error) {
        console.error(error);
        component._handleError();
      });
  }

  render() {
    return (
      <div>
        {this.state.errored && this.state.current_page === "failed" && (
          <ErrorPage prevState={this._prevState.bind(this)} />
        )}
        {!this.state.errored && this.state.current_page === "not-found" && (
          <NotFoundPage prevState={this._prevState.bind(this)} />
        )}
        {!this.state.errored && this.state.current_page === "find" && (
          <Search
            handleChange={this._handleChange.bind(this)}
            handleSelect={this._handleSelect.bind(this)}
            handleSearch={this._handleSearch.bind(this)}
            updateBy={this.state.updateBy}
          />
        )}
        {!this.state.errored && this.state.current_page === "update-form" && (
          <Form
            onError={this._handleError.bind(this)}
            nextState={this._nextState.bind(this)}
            handleClose={this._handleClose.bind(this)}
            trades={this.state.trades}
          />
        )}
        {!this.state.errored && this.state.current_page === "confirm" && (
          <div className={styles.confirmState}>
            <div>Trade Successfully Updated!</div>
            <div
              className={styles.confirm}
              onClick={this._handleClose.bind(this)}
            >
              <div className={styles.confirmTag}>Confirm</div>
            </div>
          </div>
        )}
      </div>
    );
  }
}
