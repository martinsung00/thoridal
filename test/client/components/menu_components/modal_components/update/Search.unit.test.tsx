import React from "react";
import renderer from "react-test-renderer";
import { Search } from "../../../../../../src/client/components/menu_components/modal_components/update/Search";
import { shallow, configure, ShallowWrapper } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("Modal Option Update: Modal Search Tests", function () {
  const props = {
    handleChange: jest.fn(),
    handleSelect: jest.fn(),
    handleSearch: jest.fn(),
    updateBy: "id",
  };

  afterEach(function () {
    jest.resetAllMocks();
  });

  it("should render the elements as is", function () {
    const container:
      | renderer.ReactTestRendererJSON
      | renderer.ReactTestRendererJSON[]
      | null = renderer.create(<Search {...props} />).toJSON();

    expect(container).toMatchSnapshot();
  });

  it("should render a placeholder if update by is set to date", async function (done) {
    const alternateProps = {
      handleChange: jest.fn(),
      handleSelect: jest.fn(),
      handleSearch: jest.fn(),
      updateBy: "date",
    };
    const wrapper = shallow(<Search {...alternateProps} />);

    expect(wrapper).toMatchSnapshot();

    done();
  });

  it("should call props' handle select method when a select event occurs", async function (done) {
    const event = {
      currentTarget: { value: "ticker" },
    } as React.ChangeEvent<HTMLSelectElement>;
    const wrapper: ShallowWrapper = shallow<any>(<Search {...props} />);
    const spy = jest.spyOn<any, any>(props, "handleSelect");
    const flushPromises = () => new Promise(setImmediate);

    event.preventDefault = jest.fn();

    wrapper.find("#update-by").simulate("change", event);
    await flushPromises();
    wrapper.update();

    expect(spy).toHaveBeenCalled();

    done();
  });

  it("should change update by state depending on current input value", async function (done) {
    const event = {
      target: { value: "ticker" },
    } as React.ChangeEvent<HTMLInputElement>;
    const wrapper: ShallowWrapper = shallow<any>(<Search {...props} />);
    const spy = jest.spyOn<any, any>(props, "handleChange");
    const flushPromises = () => new Promise(setImmediate);

    event.preventDefault = jest.fn();

    wrapper.find("#update-input").simulate("change", event);
    await flushPromises();
    wrapper.update();

    expect(spy).toHaveBeenCalled();

    done();
  });

  it("should call the search method upon pressing the search button", async function (done) {
    const event = {
      target: {},
    } as React.MouseEvent<HTMLDivElement, MouseEvent>;
    const wrapper: ShallowWrapper = shallow<any>(<Search {...props} />);
    const spy = jest.spyOn<any, any>(props, "handleSearch");
    const flushPromises = () => new Promise(setImmediate);

    event.preventDefault = jest.fn();

    wrapper.find("#search-button").simulate("click", event);
    await flushPromises();
    wrapper.update();

    expect(spy).toHaveBeenCalled();

    done();
  });
});
