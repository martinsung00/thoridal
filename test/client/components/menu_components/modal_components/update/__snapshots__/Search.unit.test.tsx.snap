// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Option Update: Modal Search Tests should render a placeholder if update by is set to date 1`] = `ShallowWrapper {}`;

exports[`Modal Option Update: Modal Search Tests should render the elements as is 1`] = `
<div>
  <div>
    Update Trade By:
  </div>
  <select
    defaultValue="id"
    id="update-by"
    onChange={[Function]}
    value="id"
  >
    <option
      value="id"
    >
      ID
    </option>
    <option
      value="ticker"
    >
      Ticker
    </option>
    <option
      value="company"
    >
      Company Name
    </option>
    <option
      value="date"
    >
      Date
    </option>
  </select>
  <input
    id="update-input"
    onChange={[Function]}
    placeholder=""
    required={true}
  />
  <div
    id="search-button"
    onClick={[Function]}
  >
    <div>
      Search
    </div>
  </div>
</div>
`;
